vmaas/webapp/app.py

```diff
@@ -503,105 +489,102 @@ def load_cache_to_apis():
-        else:
-            res = handler(request, **kwargs)
-
-        duration = time.time() - start_time
-        # (0)  /(1) /(2) /(3)
-        #     /api /v1  /updates
-        const_path = '/'.join(request.path.split('/')[:4])
-        REQUEST_TIME.labels(request.method, const_path).observe(duration)
-        REQUEST_COUNTS.labels(request.method, const_path, res.status).inc()
+    app = connexion.AsyncApp(__name__,
+                             swagger_ui_options=connexion.options.SwaggerUIOptions(swagger_ui=True))
 
-        return res
+    for route, spec in specs.items():
+        app.add_api(spec, resolver=connexion.RestyResolver('app'),
+                    validate_responses=False,
+                    strict_validation=False,
+                    base_path=route,
+                    pass_context_arg_name='request',
+                    arguments={'vmaas_version': VMAAS_VERSION})
 
-    @web.middleware
-    async def gzip_middleware(request, handler, **kwargs):
-        """ Middleware that compress response using gzip"""
-        res = await handler(request, **kwargs)
-        header = 'Accept-Encoding'
+    class ErrorHandlerMiddleware:
+        """Middleware to wrap error message in JSON struct."""
+        def __init__(self, app):
+            self.app = app
+            self.start_message = None
 
-        try:
-            if res.body is not None and header in request.headers and "gzip" in request.headers[header]:
-                gzipped_body = gzip.compress(res.body, compresslevel=GZIP_COMPRESS_LEVEL)
-                res.body = gzipped_body
-                res.headers["Content-Encoding"] = "gzip"
-                return res
-        except (TypeError, AttributeError):
-            LOGGER.warning("unable to gzip response '%s'", request.path)
-        return res
-
-    @web.middleware
-    async def error_handler(request, handler, **kwargs):
-        def format_error(detail, status):
+        def _format_error(self, detail):
             res = {}
             res["type"] = "about:blank"
             res["detail"] = detail
-            res["status"] = status
-
+            res["status"] = self.start_message["status"]
             return res
 
-        res = await handler(request, **kwargs)
-
-        if res.status >= 400 and res.body:
-            body = loads(res.body)
-            better_error = format_error(body, res.status)
-            return web.json_response(better_error, status=res.status)
-
-        return res
-
-    middlewares = []
+        async def __call__(self, scope, receive, send):
+            if scope["type"] != "http":
+                await self.app(scope, receive, send)
+                return
+
+            async def send_with_formatted_error(message):
+                if message["type"] == "http.response.start":
+                    self.start_message = message
+                elif message["type"] == "http.response.body":
+                    if self.start_message["status"] >= 400:
```

Comment:
is `self.start_message` set at this point, wouldn't it be an exception? I see it is set to `None` in `__init__` and then it is set only here in `__call__` but  not in this branch.

Reply:
The assumption here is that for a single response the middleware is called twice. First the `http.response.start` sections which handles headers and then the `http.response.body` section which handles body. It shouldn't happen that `self.start_message` is not set in the `http.response.body` section.